speaker/layer2_controller.go

```diff
@@ -49,20 +49,25 @@ func usableNodes(eps *v1.Endpoints, mList *memberlist.Memberlist) []string {
 	}
 
 	usable := map[string]bool{}
-	for _, subset := range eps.Subsets {
-		for _, ep := range subset.Addresses {
-			if ep.NodeName == nil {
-				continue
-			}
-			if activeNodes != nil {
-				if _, ok := activeNodes[*ep.NodeName]; !ok {
+	if isTrafficLocal {
+		for _, subset := range eps.Subsets {
+			for _, ep := range subset.Addresses {
+				if ep.NodeName == nil {
 					continue
 				}
-			}
-			if _, ok := usable[*ep.NodeName]; !ok {
-				usable[*ep.NodeName] = true
+				if activeNodes != nil {
+					if _, ok := activeNodes[*ep.NodeName]; !ok {
+						continue
+					}
+				}
+				if _, ok := usable[*ep.NodeName]; !ok {
+					usable[*ep.NodeName] = true
+				}
 			}
 		}
+	} else {
+		// For non traffic local services, all nodes are usable.
+		usable = activeNodes
```

Comment:
Hmm, if the speaker daemonset is deployed with a `nodeSelector:` then this can also fail? Let me elaborate:

Let's say 4 nodes cluster (A, B, C and D nodes), speaker ds is on nodes A and B (due to a `nodeSelector:` in the speaker ds yaml added by the user) but with this new order, nodes C and D are also a candidate. If they are elected, but the speaker doesn't run on it, it will not announce. If the reasoning is correct, though, this problem existed before the PR too (for example, not overlaping speaker pods and svc pods due to node selectors). But just trying to see ways this might fail :-)

I'm not familiar with the code and haven't even used L2 before, so this is all very theoretical. I might be missing something very obvious, sorry if that is the case :)

Reply:
The activeNodes list is built from memberlist active nodes so it will only include nodes with metalLB running. I don't see it's an issue.

Reply:
Ohh, that seems right. Thanks for noticing! :)

Reply:
sounds like this thread can be marked resolved?